sputnikdao2/src/proposals.rs

```diff
@@ -217,41 +218,218 @@ impl From<ProposalInput> for Proposal {
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()>;
+    fn callback_after_storage_balance_of(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()>;
+}
+
+#[near_bindgen]
+impl Contract {
+    #[allow(dead_code)]
+    #[private]
+    pub fn callback_after_storage_deposit(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()> {
+        assert_eq!(
+            env::promise_results_count(),
+            1,
+            "ERR_UNEXPECTED_CALLBACK_PROMISES"
+        );
+        match env::promise_result(0) {
+            PromiseResult::NotReady => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_deposit Error: Received PromiseResult::NotReady")
+            }
+            PromiseResult::Failed => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_deposit Error: Received PromiseResult::Failed")
+            }
+            PromiseResult::Successful(result) => {
+                let balance = near_sdk::serde_json::from_slice::<StorageBalance>(&result).unwrap();
+
+                // Pay back the proposal bond - registration fee.
+                Promise::new(proposer_account.clone())
+                    .transfer(attached_deposit.0 - balance.total.0);
+                self.internal_payout(&token_id, &receiver_id, amount.0, memo, msg)
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    #[private]
+    pub fn callback_after_storage_balance_of(
+        &mut self,
+        token_id: AccountId,
+        proposer_account: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+        attached_deposit: U128,
+        memo: String,
+        msg: Option<String>,
+    ) -> PromiseOrValue<()> {
+        assert_eq!(
+            env::promise_results_count(),
+            1,
+            "ERR_UNEXPECTED_CALLBACK_PROMISES"
+        );
+        match env::promise_result(0) {
+            PromiseResult::NotReady => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_balance_of Error: Received PromiseResult::NotReady")
+            }
+            PromiseResult::Failed => {
+                Promise::new(proposer_account.clone()).transfer(attached_deposit.0);
+                panic!("storage_balance_of Error: Received PromiseResult::Failed")
```

Comment:
(tagging another place, similar to another [comment](https://github.com/near-daos/sputnik-dao-contract/pull/49/files#r751128207))